### ISSUE TITLE
DSNAS: The implementation of dummy 1 gradient seems inconsistent with the formula in the paper

### ISSUE THREAD

**Issue by daisyjack** [SNAS-Series/SNAS-Series#7]
In the code, dummy 1 gradient is computed as:
```
(error_loss+loss_alpha).backward()           
self.block_reward = self.weights.grad.data.sum(-1)
```
In the code, `error_loss.backward()` will accumulate gradient dL/d(dummy 1) in `self.weights`, also `loss_apha.backward()` will accumulate extra gradient in self.weights, which is [p_0/p_k, p_1/p_k, p_2/p_k, ..., p_n/p_k] in each row of self.weights, so, `self.weight.grad.data.sum(-1)` will yield a vector, each item of the vector is dL/d(dummy 1) + 1/p_k. 
However, In the paper, the gradient should be dL/d(dummy 1). The extra gradient 1/p_k is quite large, which will affect the reward greatly.

The quoted code 
`#self.block_reward = torch.autograd.grad(error_loss, self.weights, retain_graph=True, allow_unused=True)`
seems the right way to compute dL/d(dummy 1), but it‘s incompatible with pytorch distributed settings.

Is this a bug or are there some other considerations for doing so?

**Comment by daisyjack** [comment]
oh i see. 
```
loss_alpha = torch.log((self.weights * F.softmax(self.log_alpha, dim=-1)).sum(-1)).sum()/self.args.world_size
self.weights.requires_grad_() 
```
when loss_alpha is computed, self.weights.requires_grad=False, so the computation of loss_alpha is not recorded by grad_fn
